In [22]:
class Date:
    day : int
    month : str | int
    year : int

    def __init__(
            self,
            Date : str
    ):

        import re
        from typing import Literal
        pattern = '^(\d{1,2})[-./](\d{1,2}|\w+)[-./](\d{4})$'

        compiled = re.fullmatch(pattern, Date)
        if not compiled:
                raise ValueError("Please provide the date in dd-mm-yyyy")

        day = (compiled.group(1))
        month = compiled.group(2)
        year = (compiled.group(3))
        print([day, month, year])
        if not day.isnumeric():
                raise ValueError("Day cannot be empty or a string")
        if not year.isnumeric():
                raise ValueError("Year cannot be empty or a string")

        day = int(day)
        year = int(year)
        month_Num = None

        if month.isnumeric():
                if int(month) <= 0 or int(month) >= 12:
                        raise ValueError("Not a valid Month.. Enter a month between 1 and 12")
                else:
                        month_Num = int(month)
                        month = month_Num
        if ( isinstance(month, str) ):
                if month.upper() in ["JAN", "JANUARY"]:
                        month_Num = 1
        
                elif month.upper() in ["FEB", "FEBRUARY"]:
                        month_Num = 2

                elif month.upper() in ["MAR", "MARCH"]:
                        month_Num = 3

                elif month.upper() in ["APR", "APRIL"]:
                        month_Num = 4
                
                elif month.upper() in ["MAY"]:
                        month_Num = 5
                
                elif month.upper() in ["JUN", "JUNE"]:
                        month_Num = 6

                elif month.upper() in ["JUL", "JULY"]:
                        month_Num = 7

                elif month.upper() in ["AUG", "AUGUST"]:
                        month_Num = 8

                elif month.upper() in ["SEP", "SEPTEMBER"]:
                        month_Num = 9

                elif month.upper() in ["OCT", "OCTOBER"]:
                        month_Num = 10

                elif month.upper() in ["NOV", "NOVEMBER"]:
                        month_Num = 11

                elif month.upper() in ["DEC", "DECEMBER"]:
                        month_Num = 12
                else:
                        raise ValueError("Not a valid Month")


        daysInAMonthDict = {
        1 : 30,
        2 : 28,
        3 : 31,
        4 : 30,
        5 : 31,
        6 : 30,
        7 : 31,
        8 : 31,
        9 : 30,
        10 : 31,
        11 : 30,
        12  : 31
        }

        def isLeapYear(
                year : int
        ) -> bool:
                if year % 100 == 0:
                        if year % 400 == 0:
                                return True
                        else:
                                return False
                else:
                        if year % 4 == 0 :
                                return True
                        else:
                                return False

        if isLeapYear(year):
                daysInAMonthDict[2] = 29

        if day <= 0 or day > daysInAMonthDict[month_Num]:
                raise ValueError("Invalid number of days") 


        self.day = day
        self.month = month_Num
        self.year = year

In [15]:
from datetime import datetime

dateNow = datetime.now()
year = dateNow.strftime("%b %d-%m-%Y %H:%M:%S")
year

'Jun 18-06-2025 12:41:51'

In [56]:
from itertools import permutations

possibleFormats = []
d = ["%d"]
m = ["%b", "%B", "%m"]
y = ["%y", "%Y"]
delimiter = [".", "/", " ", "-"]

mapping = {
    'd': d,
    'm': m,
    'y': y,
}

all_permutations = list(permutations(['d', 'm', 'y']))

for sep in delimiter: 
    for perm in all_permutations:
        for x in mapping[perm[0]]:
            for y_ in mapping[perm[1]]:
                for z in mapping[perm[2]]:
                    s = f"{x}{sep}{y_}{sep}{z}"
                    possibleFormats.append(s)
print(len(possibleFormats))


144


In [106]:
import re
from datetime import datetime

def remove_day_suffixes(date_str: str) -> str:
    return re.sub('(\d+)\s*(st|nd|rd|th)', r'\1', date_str)

def normalize_date(date_str: str) -> str:
    date_str = date_str.strip()                    
    date_str = remove_day_suffixes(date_str)
    date_str = date_str.replace(',', '')           
    date_str = date_str.replace('.', '-')          
    date_str = re.sub(r'\s+', ' ', date_str) 
    print("BEFORE REGEX : " + date_str)   
    # date_str = re.sub(r'\s*(\d{1,2})\s*[/-]\s*(\d{1,2}|\w+)\s*[/-]\s*(\d{2,4})', r'\1/\2/\3', date_str)
    date_str = date_str.replace(" ", "")
    return date_str

def standardize_date(date_str: str) -> str:
    date_str = normalize_date(date_str)
    print("Normalized date_str = " + date_str)

    # knownFormats = [
    #     "%d-%m-%Y", 
    #     "%d-%m-%y",
    #     "%d/%m/%Y", "%d/%m/%y",
    #     "%Y-%m-%d", 
    #     "%Y/%m/%d",
    #     "%d-%b-%Y", 
    #     "%d-%B-%Y",
    #     "%d %b %Y", "%d %B %Y",
    #     "%b %d %Y",
    #       "%B %d %Y",
    #     "%b %d, %Y", "%B %d, %Y",
    #     "%b %d %y", 
    #     "%B %d %y",
    #     "%Y %b %d", "%Y %B %d",
    #     "%b-%d-%Y", "%B-%d-%Y",
    #     "%b-%d-%y", "%B-%d-%y",
    #     "%d.%m.%Y", "%d.%m.%y",
    #     "%d %B %y", "%d/%B/%y", "%d/%B/%Y", "%d/%b/%y", "%d/%b/%y", "%y/%b/%d", "%y/%B/%d"
    # ]
    for fmt in possibleFormats:
        try:
            dt = datetime.strptime(date_str, fmt)
            return dt.strftime("%d-%m-%Y")
        except ValueError:
            continue

    raise ValueError(f"Unsupported date format: {date_str}")


In [119]:
standardize_date("April- 25  -24")

BEFORE REGEX : April- 25 -24
Normalized date_str = April-25-24


'25-04-2024'

In [98]:
standardize_date("37/ February/ 21")

BEFORE REGEX : 37/ February/ 21
Normalized date_str = 37/February/21


'21-02-2037'

In [ ]:
while True:
    print(standardize_date(input()))

After removing suffixes 5 June 37
Normalized date_str = 5 June 37
05-06-2037
After removing suffixes 7 mAY 99
Normalized date_str = 7 mAY 99
07-05-1999
After removing suffixes 09/09/12
Normalized date_str = 09/09/12
09-09-2012


Year can be at the starting or at the end
Year can be like [ " '95 ", " 1995 " ]

Month can be only at the first or in the middle
Month can be represented by [ 9, Sep, September ]

Day is represented only as %d

The delimiter used are [ " ", /, -, . ]

__ (delimeter) ____ (delimeter) ____
[ dmy, dym, mdy, myd, ymd, ydm ]
d-m-y
1 * 3 * 2 * 4 = 24

m d y
3 * 1 * 2 * 4 = 24

...
There are 6 Permutations 
6 * 24 = 144

In [50]:
from itertools import permutations
my_list = ["d", "m", "y"]
all_permutations = list(permutations(my_list))
for i in all_permutations:
    print(i)

('d', 'm', 'y')
('d', 'y', 'm')
('m', 'd', 'y')
('m', 'y', 'd')
('y', 'd', 'm')
('y', 'm', 'd')


144